### PR TITLE
chore: fetch remote reference in local

### PR DESCRIPTION
**Pull request by sondermanish**
## Description
- Added fetch remote CGS Impl

Fixes #
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Git"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12233198064>
> Commit: d964d59e0347f1ad8592edbdf3158818c7f26a17
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12233198064&attempt=2" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Git`
> Spec:
> <hr>Mon, 09 Dec 2024 20:08:50 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Release Notes

- **New Features**
	- Introduced new methods for fetching remote changes in various Git services, enhancing synchronization capabilities.
	- Added new fields for internal metadata management.
	- Implemented user session management improvements for better handling of user-related data.
	- Added analytics tracking for unit execution time to improve performance insights.

- **Bug Fixes**
	- Enhanced error handling in new fetch methods to ensure robustness.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCE.java [hunk 1]
```diff
@@ -23,4 +24,14 @@ Mono<String> commitArtifact(
...
     Mono<? extends Artifact> detachRemote(String branchedArtifactId, ArtifactType artifactType, GitType gitType);
+
+    Mono<String> fetchRemoteChanges(
+            String referenceArtifactId,
+            boolean isFileLock,
+            ArtifactType artifactType,
+            GitType gitType,
+            RefType refType);
+
+    Mono<String> fetchRemoteChanges(
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCE.java:35]: Why are these methods getting exposed on the interface twice?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 2]
```diff
@@ -67,6 +70,7 @@
 import static com.appsmith.external.helpers.AppsmithBeanUtils.copyNestedNonNullProperties;
 import static com.appsmith.server.constants.FieldName.DEFAULT;
 import static com.appsmith.server.constants.SerialiseArtifactObjective.VERSION_CONTROL;
+import static com.appsmith.server.constants.ce.FieldNameCE.BRANCH_NAME;
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:73]: Gotta create a lint rule for this I think 😶‍🌫️

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 3]
```diff
@@ -940,6 +945,108 @@ private boolean isBaseGitMetadataInvalid(GitArtifactMetadata gitArtifactMetadata
...
+            return Mono.error(new AppsmithException(AppsmithError.GIT_GENERIC_ERROR));
+        }
+
+        GitArtifactMetadata baseArtifactGitData = baseArtifact.getGitArtifactMetadata();
+        GitArtifactMetadata refArtifactGitData = refArtifact.getGitArtifactMetadata();
+
+        String baseArtifactId = baseArtifactGitData.getDefaultArtifactId();
+
+        if (refArtifactGitData == null || !hasText(refArtifactGitData.getRefName())) {
+            return Mono.error(new AppsmithException(AppsmithError.INVALID_PARAMETER, BRANCH_NAME));
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:964]: TODO for error message accepting git type

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 4]
```diff
@@ -940,6 +945,108 @@ private boolean isBaseGitMetadataInvalid(GitArtifactMetadata gitArtifactMetadata
...
+        }
+
+        Mono<User> currUserMono = sessionUserService.getCurrentUser().cache(); // will be used to send analytics event
+        Mono<Boolean> acquireGitLockMono =
+                gitRedisUtils.acquireGitLock(baseArtifactId, GitConstants.GitCommandConstants.FETCH_REMOTE, isFileLock);
+
+        ArtifactJsonTransformationDTO jsonTransformationDTO = new ArtifactJsonTransformationDTO();
+        jsonTransformationDTO.setRefType(RefType.BRANCH);
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:972]: Duplicate setter

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 5]
```diff
@@ -940,6 +945,108 @@ private boolean isBaseGitMetadataInvalid(GitArtifactMetadata gitArtifactMetadata
...
+                        gitHandlingService.fetchRemoteChanges(jsonTransformationDTO, baseArtifactGitData.getGitAuth())))
+                .flatMap(fetchedRemoteStatusString -> {
+                    return gitRedisUtils.releaseFileLock(baseArtifactId).thenReturn(fetchedRemoteStatusString);
+                })
+                .onErrorResume(throwable -> {
+                    /*
+                     in case of any error, the global exception handler will release the lock
+                     hence we don't need to do that manually
+                    */
+                    log.error(
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:994]: Error can already include git type

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java (hunk 4 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:972]: resolved it.

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**